libyul/optimiser/DeadCodeEliminator.cpp

```diff
@@ -0,0 +1,134 @@
+/*(
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Optimisation stage that removes unreachable code.
+ */
+
+#include <libyul/optimiser/DeadCodeEliminator.h>
+#include <libyul/AsmData.h>
+
+#include <libevmasm/SemanticInformation.h>
+#include <libevmasm/AssemblyItem.h>
+
+#include <algorithm>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+using SemInfo = eth::SemanticInformation;
+
+namespace
+{
+bool isTerminating(yul::ExpressionStatement const& _exprStmnt)
+{
+	if (_exprStmnt.expression.type() != typeid(FunctionalInstruction))
+		return false;
+
+	auto funcInstr = boost::get<FunctionalInstruction>(_exprStmnt.expression);
+
+	return SemInfo::terminatesControlFlow(eth::AssemblyItem{funcInstr.instruction});
+}
+
+// Returns an iterator to the first terminating statement or ::end() when none
+// was found
+template <typename B>
+auto findDeadCode(B& _block)
+{
+	auto& stmnts = _block.statements;
+
+	return find_if(
+		stmnts.begin(),
+		stmnts.end(),
+		[](Statement const& _stmnt)
+		{
+			if (_stmnt.type() == typeid(ExpressionStatement))
+			{
+				auto exprStmnt = boost::get<ExpressionStatement>(_stmnt);
+
+				if (isTerminating(exprStmnt))
+					return true;
+			}
+
+			if (_stmnt.type() == typeid(yul::Instruction))
+			{
+				auto instr = boost::get<yul::Instruction>(_stmnt);
+
+				if (SemInfo::terminatesControlFlow(eth::AssemblyItem{instr.instruction}))
+					return true;
+			}
+
+			if (_stmnt.type() == typeid(Break))
+				return true;
+
+			if (_stmnt.type() == typeid(Continue))
+				return true;
+
+			return false;
+		}
+	);
+}
+}
+
+bool DeadCodeEliminator::containsDeadCode(Block const& _block)
+{
+	m_found_dead_code = false;
+
+	(*this)(_block);
+
+	return m_found_dead_code;
+}
+
+void DeadCodeEliminator::operator()(Block const& _block)
+{
+	auto result = findDeadCode(_block);
+
+	if (result != _block.statements.cend() &&
+		result+1 != _block.statements.cend())
+	{
+		m_found_dead_code = true;
+		return;
+	}
+
+	ASTWalker::operator()(_block);
+}
+
+void DeadCodeEliminator::operator()(Block& _block)
+{
+	auto& stmnts = _block.statements;
+
+	auto result = findDeadCode(_block);
+
+	if (
+		result != stmnts.end() &&
+		result+1 != stmnts.end()
+	)
+		stmnts.erase(
```

Comment:
Could use `boost::remove_erase_if`

Reply:
I only want to start removing after the terminating statement, as far as I could see, `boost::remove_erase_if` only takes the whole container as argument?

Reply:
Ah right!